### PR TITLE
Reconstruct new electrons with H/E (cone) > 0.15 and H/E (tower) < 0.15

### PR DESCRIPTION
**Pull request by brentstone**
We are proposing to reconstruct new ecalDriven electrons with H/E (cone) > 0.15 and H/E (tower behind SC) < 0.15. 

We spoke on 11/30 to the RECO group regarding these changes: https://indico.cern.ch/event/776884/contributions/3232028/attachments/1763017/2861031/EgammaReco_11_30_18.pdf.

The issues mentioned in the talk have been resolved with a nice simple solution. No original electrons are lost; only new ones are gained.

As mentioned in the talk linked above, this change was tested on three samples: a JetHT data sample, Egamma data, and ttbar RelVal MC. The file size increases are marginal. JetHT increases by 0.6% in MINIAOD and 0.03% in AOD. Egamma increases by 0.07% in MINIAOD and 0.02% in AOD, while sometimes fluctuating to - 0.03%. Ttbar fluctuates as well, experiencing 0.2% increases to -0.005% decreases in both MINIAOD and AOD.

In 1000 events, we gain 99 electrons in JetHT, 18 electrons in Egamma, and 46 electrons in ttbar.

The timeReports show that the timing increase by including these changes is marginal. The text files of these reports are provided as attached files.
[timeReport_TTBAR.txt](https://github.com/cms-sw/cmssw/files/2655948/timeReport_TTBAR.txt)
[timeReport_JETHT.txt](https://github.com/cms-sw/cmssw/files/2655949/timeReport_JETHT.txt)
[timeReport_EGAMMA.txt](https://github.com/cms-sw/cmssw/files/2655950/timeReport_EGAMMA.txt)

The changes have also been tested on a 1000 event sample of single-electron Radion->HH->bbWW->e nu qq. At 1 TeV, the electron reconstruction efficiency increases from 65% to 91%. At 3 TeV, the increase is from 33% to 73%. This change has the potential to improve boosted analyses with leptons and jets by quite a bit.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@brentstone @Sam-Harper 
do I understand correctly that this PR is not yet ready for review of its physics impact?
Please clarify.
Thank you.

File: RecoEgamma/EgammaElectronAlgos/src/GsfElectronAlgo.cc [hunk 1]
```diff
@@ -1114,17 +1114,15 @@ bool GsfElectronAlgo::isPreselected( GsfElectron * ele )
  {
 	bool passCutBased=ele->passingCutBasedPreselection();
 	bool passPF=ele->passingPflowPreselection(); //it is worth nothing for gedGsfElectrons, this does nothing as its not set till GedGsfElectron finaliser, this is always false
-	if(generalData_->strategyCfg.gedElectronMode){
+        if(generalData_->strategyCfg.gedElectronMode){
```
Comment (slava77) [inline RecoEgamma/EgammaElectronAlgos/src/GsfElectronAlgo.cc]: indentation

Comment (slava77) [reply RecoEgamma/EgammaElectronAlgos/src/GsfElectronAlgo.cc]: (you may need to untabify)
usually a local check with ```git diff ``` in the terminal is enough to spot these inconsistencies before pushing to the remote

File: RecoEgamma/EgammaElectronAlgos/src/GsfElectronAlgo.cc [hunk 2]
```diff
@@ -1181,12 +1179,14 @@ void GsfElectronAlgo::setCutBasedPreselectionFlag( GsfElectron * ele, const reco
...
+  double hoeCone = ele->hcalOverEcal();
+  double hoeTower = ele->hcalOverEcalBc();
   const reco::CaloCluster & seedCluster = *(ele->superCluster()->seed()) ;
   int detector = seedCluster.hitsAndFractions()[0].first.subdetId() ;
   bool HoEveto = false ;
-  if (detector==EcalBarrel && (had<cfg->maxHBarrel || (had/ele->superCluster()->energy())<cfg->maxHOverEBarrel)) HoEveto=true;
-  else if (detector==EcalEndcap && (had<cfg->maxHEndcaps || (had/ele->superCluster()->energy())<cfg->maxHOverEEndcaps)) HoEveto=true;
+  double scle = ele->superCluster()->energy();
+  if (detector==EcalBarrel && (hoeCone*scle<cfg->maxHBarrelCone || hoeTower*scle<cfg->maxHBarrelTower || hoeCone<cfg->maxHOverEBarrelCone || hoeTower<cfg->maxHOverEBarrelTower)) HoEveto=true;
+  else if (detector==EcalEndcap && (hoeCone*scle<cfg->maxHEndcapsCone || hoeTower*scle<cfg->maxHEndcapsTower || hoeCone<cfg->maxHOverEEndcapsCone || hoeTower<cfg->maxHOverEEndcapsTower)) HoEveto=true;
```
Comment (slava77) [inline RecoEgamma/EgammaElectronAlgos/src/GsfElectronAlgo.cc]: please add some newlines for readability

perhaps a better option is to rewrite a bit like
```c++
if (detector == EcalBarrel) HoEveto = 
     hoeCone*scle<cfg->maxHBarrelCone || hoeTower*scle<cfg->maxHBarrelTower || 
    hoeCone<cfg->maxHOverEBarrelCone || hoeTower<cfg->maxHOverEBarrelTower;
else if (detector == EcalEndcap) ...

```

**Review by slava77 — COMMENTED** [review]
mostly trivialities to fixup

File: RecoEgamma/EgammaElectronAlgos/src/GsfElectronAlgo.cc (hunk 1 above)
Comment (brentstone) [reply RecoEgamma/EgammaElectronAlgos/src/GsfElectronAlgo.cc]: thx, i have fixed this

**Comment by Sam-Harper** [comment]
>do I understand correctly that this PR is not yet ready for review of its physics impact?
>Please clarify.

Yes after discussion with JetMET, we need to do one more thing which is to remove these new electrons from particle flow. We wanted to show  them our proposed changes first via a PR but they have requested that change.  Tau were okay to proceed with the PR as is. I mean it probably removes one more variable when trying to figure out what to do for the ultra legacy so I can see the reasoning

**Comment by brentstone** [comment]
Merged some changes from Sam Harper which exclude the new electrons from Particle Flow

**Comment by Sam-Harper** [comment]
thanks @brentstone.  So this PR is now ready. It should add gedGsfElectrons to the event but should not alter the number of particle flow candidates. So far a check of 1000 JetHT events shows no change to the number of PFCandidates in the event with new commit.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by perrotta** [comment]
@brentstone @Sam-Harper 
Looking at the comparisons in the automatic tests, I see that puppi Jets and MET are modified in miniAOD by this PR: does it still satisfy the agreement you reached with the JetMET group?

**Comment by perrotta** [comment]
@Sam-Harper "failing" comparisons can be seen all together under "See failed" in the test results linked in https://github.com/cms-sw/cmssw/pull/25440#issuecomment-446924265. Any of the miniAOD comparisons there shows the differences in puppi jet/met.
(Maybe they also look at gsf tracks, not only pf candidates: I don't remember...)

**Comment by perrotta** [comment]
@brentstone @Sam-Harper : did you have any opportunity to conclude your investigations in the past days?

**Comment by perrotta** [comment]
This PR needs to be rebased

**Comment by Sam-Harper** [comment]
okay, expect movement soon on this, triaging with the pf rec-hit threholds and other 10_5 mandatory features.

**Comment by ahinzmann** [comment]
We would expect PUPPI jets to stay the same, but PUPPI isolation and PUPPI MET to change.
PUPPI jets are clustered from all PF candidates (puppiWeight).
PUPPI isolation (puppiWeightNoLep) uses all PF candidates that are not electrons (by pdg ID).
PUPPI MET uses both puppiWeight and puppiWeightNoLep.
Is this the case. I can't access the plots via the link above somehow.

**Comment by perrotta** [comment]
Thank you @ahinzmann !
You can't access the link because it gets emptied after some time (tests were run more than one month ago...). Please solve the  conflicts reported and commit the fix, so that tests can be re-run and you can still access their output

**Comment by perrotta** [comment]
@brentstone please rebase

**Comment by Sam-Harper** [comment]
hi @brentstone , I have an example of the rebase here: https://github.com/cms-sw/cmssw/compare/master...Sam-Harper:feature_ElectronReco_TowerHoE_105X-20190115?expand=1

You can just copy that branch to your area and resubmit a PR or force update. If I dont hear from you tonight, I'll open up a PR myself just to make sure it makes it

**Comment by brentstone** [comment]
hi @Sam-Harper,
I’ve been trying to do a rebase but am confused by the history of the commits here.

After fetching and creating a new branch, I’ve tried something like the command:

git rebase —onto CMSSW_10_5_X HEAD~3 feature_ElectronReco_HcalTowerReco_105X

However this gave me a fatal error. I am not sure if CMSSW_10_5_X is the branch I want to rebase off of or if it is another variant of CMSSW_10_5. I also figured I should do ~3 because our PR has 3 total commits now (two from me, one from you).

Looking at the merge conflicts as they currently exist now, it seems some of the files have changed quite a bit as well.

If you could offer some clearer guidance on this, I’d be happy to figure it out.

**Comment by Sam-Harper** [comment]
the branch needed is a nightly so like CMSSW_10_5_X_2019-01-15-1100

git checkout brentstone/feature_ElectronReco_HcalTowerReco_1040pre1
git rebase --onto CMSSW_10_5_X_2019-01-15-1100 HEAD~4 HEAD
git checkout --ours RecoEgamma/EgammaElectronProducers/plugins/GsfElectronBaseProducer.cc
git checkout --ours RecoEgamma/EgammaElectronProducers/python/gedGsfElectrons_cfi.py
git checkout --ours RecoEgamma/EgammaElectronProducers/python/gsfElectrons_cfi.py
(I always get confused between --ours and theirs, here its --ours I beleive)
Then see the changes I made to GsfElectronBaseProducer in that branch I linked and impliment them. Otherwise use my branch.

**Comment by perrotta** [comment]
-1
@brentstone please close this PR, as it is now superseded by #25699 
By the way, I notice that having added a parameter set validation in the GsfElectronProducers with #25574 (which got included in the rebase made to produce #25699) allowed to spot the wrong HI configuration that was fixed with https://github.com/cms-sw/cmssw/pull/25699/commits/29c65f48a04a7dab841e7c6756fb94530b8b9da4 , which was instead overlooked in this older version without that parameter set validation

**Comment by Sam-Harper** [comment]
Indeed this highlights the importance of the pset validation and also Slava's favourite option of overriding parameters as "mod.param = 3.0" rather than "mod.param = cms.double(3.0)"

**Pull request closed without merging**